Modernize flesctl

cuveland opened the pull request:
Rewrites flesctl in Python with a YAML configuration and adapts it to the current `stserver`/`tsmanager`/`tsbuilder` chain. Installs as `flesctl-ng` next to the legacy shell version, which is left untouched.

- `flescfg.py` — configuration format, merging of tags, validation
- `init_run.py` — generates one configuration file per role plus the SPM job
- `fles_input.py` — configures the CRIs of a node and starts its stserver
- `flesctl.py` — command line interface
- `test/run-local` — runs all roles on one machine and checks the result

The roles find each other through the tsmanager and retry until it answers, so no startup order is enforced and the SPM semaphores are gone.

### Tested

A local run with the software pattern generator: 20 timeslices through all four roles, analyzer clean, recorded archive read back intact. Generated configuration files are accepted by all three binaries.

### Not tested

Never run on the cluster, so the multi-node case, `ucx_net_devices` and `advertise-host` are unverified. `flesctl.py` itself has not been executed.

### Known issues

- `tsmanager` and `stserver` exit with code 0 when creating their UCX listener fails, so a failed start is not detectable. They also cannot restart for about a minute, as TIME-WAIT blocks the listener. Together a too-early restart yields a silently empty run.
- The legacy `contrib/flesctl` cannot work against current master and should be removed separately, along with deciding whether the new one takes over the `flesctl` name.

cuveland committed:
Use a pager for config output in flesctl

cuveland committed:
Add example configuration file for flesctl with en03_pgen settings

cuveland committed:
Fix invalid filename handling in flesctl

cuveland committed:
Add flescfg test script

cuveland committed:
Refactor flescfg into flescfg.py with YAML merging and validation

cuveland committed:
Add init_run.py Python script for generating Flesnet and SPM configurations

cuveland committed:
Add comparison script for shell and Python versions of Flesnet configurations

cuveland committed:
Add fles_input.py for managing Flesnet input processes using Python

PALoizeau commented on contrib/flesctl/init_run.py:
```diff
@@ -0,0 +1,124 @@
+                    input += 1
+            EN_CFG.append("")
+            CRI_SPM.append(f"cri_{node}_{cardinfo['pci_address'][:2]}")
+        EN_SPM.append(f"{node}: bash -c \"LOGDIR={LOGDIR}{node}_ {FLESCTLDIR}/fles_input {config_file} {node}\"")
```
```suggestion
        EN_SPM.append(f"{node}: bash -c \"LOGDIR={LOGDIR}{node}_ {FLESCTLDIR}/fles_input.py {config_file} {node}\"")
```
Hi Jan

If I understood your scripts and what I saw in the final diff of the compare tool correctly, the Python version of the input script should be used if the input is yaml, is that right?

PALoizeau commented:
Really nice proposal :+1: 
I had started experimenting on something similar on the ECS side while thinking about how to handle the config tag to parameters sequence, and while I ended with a YAML format similar to yours (same reference config after all), I really prefer yours
Especially the node specific configuration is more flexible in what you did compared to my attempts.

One thing that came to my mind while reading you PR is that we could use this opportunity to prepare support for the following two points while keeping a "legacy" output for now:
- More flexibility in where the 3 parts of the DAQ chain are run (Input, TS building and TS "usage") by introducing a list of "build nodes"
  => For the time being, if the same nodes are used as build and processing node + the list of SPM entries is reordered, it should the same as the previous EN/PN dichotomy
  => I think this would allow to prepare for having the TS building on the EN, as well as having eventual "RAW archiving" dedicated nodes (any combination from a single node to 3 node layers should then work)
- Separate `active` and `recording` states by having separate entries for the analyzer, publisher and recorder `tsclients` in the YAML (formally, as probably cannot prevent users  from misusing them (^^)' )
  => For the time being, if the entries in the 3 lists are merged before the loop on tsclients, it should lead to the same SPM footprint
  => Later, we could have 
  - the `analyzers` started during the `configuring` state and used for validating some number of TS, 
  - the `publishers` started when transitioning from `configuring` to `active` once everything is OK, 
  - the `recorders` started/stopped when transitioning from `active` to `recording` (e.g. after detectors have run monitors for some time) and backward

What do you think, would this make sense?
If yes, I started playing for 1-2 hours this morning with your PR and got the following patch which seems to work for me, addressing only the "for the time being" and "same as legacy" parts as I did it just to check if it was possible at all :sweat_smile: 
(also introduced support for "recording" tsclients in the process)
[flesnet_pr175_future_yaml_prep.patch.log](https://github.com/user-attachments/files/18164614/flesnet_pr175_future_yaml_prep.patch.log)


In any case I will keep an eye on this PR and plan the SCA so it can later import this directly and make use of a pair `mFlesParServer/mFlesParSet` using this YAML format

**PS:** Had to add a ".log" to the name of my patch file for some reason, Github was refusing to accept it :confused:

cuveland committed:
Update contrib/flesctl/init_run.py

Co-authored-by: PALoizeau <p.-a.loizeau@gsi.de>

cuveland committed:
Update validation for channels in flescfg.py

cuveland committed:
Add dummy command scripts for flesctl development

cuveland committed:
Update example configuration files for flesctl

cuveland committed:
Fix output redirection in fles_input script

cuveland committed:
Refactor subprocess management in fles_input.py

cuveland committed:
Add .gitignore and update compare-it script

cuveland committed:
Refactor fles_input.py

cuveland committed:
Fix a typo

cuveland committed:
Add flesctl-ng script

cuveland committed:
Apply several fixes to flesctl scripts

cuveland committed:
Add flesctl-ng to build system

cuveland committed:
Work on flescfg structure

cuveland committed:
Refactor init_run.py

cuveland committed:
Move new flesctl to new toplevel directory

cuveland committed:
Rename flesctl-ng

cuveland committed:
Use direct function call to init_run.main

cuveland committed:
Make fles_input.py importable

cuveland committed:
Apply some minor fixes

cuveland committed:
Improve python coding standard

cuveland committed:
Refactor code style

cuveland committed:
Refactor fles_input.py

cuveland committed:
Enhance documentation in flescfg.py and init_run.py

cuveland committed:
Suppress type checking warnings in flescfg.py and flesctl.py

cuveland committed:
Introduce slurm utilities module

cuveland committed:
Remove logbook command from flesctl.py

cuveland committed:
Remove sinfo command from flesctl.py

cuveland committed:
Remove configuration management commands from flesctl.py

cuveland committed:
Refactor flesctl.py

cuveland committed:
Migrate from docopt to argparse in flesctl

cuveland committed:
Add type hints and docstrings to functions in flesctl.py

cuveland committed:
Implement GlobalConfig class for managing configuration and run IDs

cuveland committed:
Add a few type hints

cuveland committed:
Add descriptions for commands

cuveland committed:
Add support for multiple tags in flesctl

cuveland committed:
Refactor configuration to include 'active' field for entry and build nodes

cuveland committed:
Refactor command line tag handling

cuveland committed:
Add flesctl directory to the build system

The new Python-based flesctl was added as a top-level directory, but
never registered with add_subdirectory(), so its install rules were
never evaluated and the flesctl-ng component was not installed.

Also install slurm.py, which was missing from the install list.

cuveland committed:
Adapt flesctl to the stserver/tsmanager/tsbuilder chain

The legacy flesnet and cri_server binaries the generator targeted no
longer exist. Generate one configuration file per role instead of
flesnet.cfg, and group the settings accordingly.

The roles find each other through the tsmanager and retry until it
answers, so the SPM semaphores are gone; the only ordering left is
cri_cfg before stserver, sequential within fles_input.py.

Timeslice consumers are split into analyzers, publishers and recorders
so that they can later follow the ECS state transitions individually.

cuveland committed:
Add a local test run for the generated configuration

Starts all four roles on one machine, taking the data from the software
pattern generator of the stserver instead of a CRI and driving them from
the generated job description, so that what flesctl produces can be
exercised without hardware.

An entry node without cards is now valid and skips the CRI setup, which
is what makes such a run possible.

cuveland committed:
Abort the readout if configuring a CRI fails

The return code of cri_cfg was ignored. As the stserver determines the
set of enabled channels once at startup, a failed configuration would
have left the run going with the wrong channels and no indication that
anything was wrong.

cuveland committed:
Simplify the node-local readout agent

Only a single stserver is started per node now, so the machinery for a
list of subprocesses is no longer needed. Its polling loop also ended
the script successfully when the stserver died; the exit code is now
passed on instead.

Drop the wait before enabling the pattern generators, as the data path
is established asynchronously, and remove the unused slurm module.

Pass the stserver configuration file explicitly rather than relying on
the working directory.

cuveland committed:
Shut down on a signal without deferring

The guard delayed the shutdown until the readout was fully started. It
dated from the time the script waited for a number of cri_servers to
come up; what remains before the stserver runs is the configuration of
the CRIs, during which a shutdown needs no special treatment.

cuveland committed:
Check the recorded data in the local test run

Add a publisher and a recorder to the test configuration, so that all
three groups of clients and the substitution in their parameters are
covered, and read the written archive back afterwards. The analyzer
reports pattern errors in its log but still exits successfully, so the
log is checked as well.

Wait for the ports to be released before starting, as the roles cannot
listen while a TIME-WAIT entry of a preceding run exists.

cuveland committed:
Drop the scaffolding for comparing against the shell version

The dummy commands and compare-it existed to check the Python output
against the shell scripts, which no longer run against current master.
